File: backend/app/core/ai_usage_limits.py
"""Central plan limits for AI feature usage (feature-based, not tokens)."""

from __future__ import annotations

from typing import Literal

from app.models.enums import Plan

AiFeature = Literal[
    "chat",
    "quiz",
    "assignment",
    "pdf_analysis",
    "summary",
    "study_plan",
]

# None = unlimited (fair-usage; still tracked for monitoring)
PlanLimits = dict[str, int | None]

PLAN_LIMITS: dict[str, PlanLimits] = {
    Plan.FREE.value: {
        "daily_chat": 25,
        "monthly_chat": 600,
        "quiz": 10,
        "assignment": 5,
        "pdf_analysis": 3,
        "summary": 10,
        "study_plan": 2,
    },
    Plan.PRO.value: {
        "daily_chat": None,
        "monthly_chat": None,
        "quiz": None,
        "assignment": None,
        "pdf_analysis": None,
        "summary": None,
        "study_plan": None,
    },
    Plan.PREMIUM_PLUS.value: {
        "daily_chat": None,
        "monthly_chat": None,
        "quiz": None,
        "assignment": None,
        "pdf_analysis": None,
        "summary": None,
        "study_plan": None,
    },
    Plan.ENTERPRISE.value: {
        "daily_chat": None,
        "monthly_chat": None,
        "quiz": None,
        "assignment": None,
        "pdf_analysis": None,
        "summary": None,
        "study_plan": None,
    },
}

FEATURE_LABELS: dict[AiFeature, str] = {
    "chat": "AI Chat",
    "quiz": "Quiz Generation",
    "assignment": "Assignment Generation",
    "pdf_analysis": "PDF Analysis",
    "summary": "AI Note Summary",
    "study_plan": "Study Plan Generation",
}

FEATURE_TO_LIMIT_KEY: dict[AiFeature, str] = {
    "chat": "monthly_chat",  # chat also checks daily separately
    "quiz": "quiz",
    "assignment": "assignment",
    "pdf_analysis": "pdf_analysis",
    "summary": "summary",
    "study_plan": "study_plan",
}


def get_plan_limits(plan: str | Plan) -> PlanLimits:
    key = plan.value if isinstance(plan, Plan) else str(plan).upper()
    return PLAN_LIMITS.get(key, PLAN_LIMITS[Plan.FREE.value])


def is_unlimited(limit: int | None) -> bool:
    return limit is None

File: src/lib/ai-usage-limits.ts
/**
 * Central plan limits for AI feature usage (feature-based, not tokens).
 * Keep in sync with backend/app/core/ai_usage_limits.py
 */

import type { UserPlan } from "@/types/user";

export type AiFeature =
  | "chat"
  | "quiz"
  | "assignment"
  | "pdf_analysis"
  | "summary"
  | "study_plan";

export type PlanLimits = {
  daily_chat: number | null;
  monthly_chat: number | null;
  quiz: number | null;
  assignment: number | null;
  pdf_analysis: number | null;
  summary: number | null;
  study_plan: number | null;
};

export const PLAN_LIMITS: Record<UserPlan, PlanLimits> = {
  FREE: {
    daily_chat: 25,
    monthly_chat: 600,
    quiz: 10,
    assignment: 5,
    pdf_analysis: 3,
    summary: 10,
    study_plan: 2,
  },
  PRO: {
    daily_chat: null,
    monthly_chat: null,
    quiz: null,
    assignment: null,
    pdf_analysis: null,
    summary: null,
    study_plan: null,
  },
  PREMIUM_PLUS: {
    daily_chat: null,
    monthly_chat: null,
    quiz: null,
    assignment: null,
    pdf_analysis: null,
    summary: null,
    study_plan: null,
  },
  ENTERPRISE: {
    daily_chat: null,
    monthly_chat: null,
    quiz: null,
    assignment: null,
    pdf_analysis: null,
    summary: null,
    study_plan: null,
  },
};

export const FEATURE_LABELS: Record<AiFeature, string> = {
  chat: "AI Chat",
  quiz: "Quiz Generation",
  assignment: "Assignment Generation",
  pdf_analysis: "PDF Analysis",
  summary: "AI Note Summary",
  study_plan: "Study Plan Generation",
};

export const USAGE_COMPARISON_ROWS: {
  feature: string;
  free: string;
  pro: string;
  premium: string;
}[] = [
  {
    feature: "AI Chats",
    free: "25/day, 600/month",
    pro: "Unlimited",
    premium: "Unlimited",
  },
  {
    feature: "Quiz Generation",
    free: "10/month",
    pro: "Unlimited",
    premium: "Unlimited",
  },
  {
    feature: "Assignments",
    free: "5/month",
    pro: "Unlimited",
    premium: "Unlimited",
  },
  {
    feature: "PDF Analysis",
    free: "3/month",
    pro: "Unlimited",
    premium: "Unlimited",
  },
  {
    feature: "AI Summaries",
    free: "10/month",
    pro: "Unlimited",
    premium: "Unlimited",
  },
  {
    feature: "Study Plans",
    free: "2/month",
    pro: "Unlimited",
    premium: "Unlimited",
  },
];

export function getPlanLimits(plan: string): PlanLimits {
  const key = plan.toUpperCase() as UserPlan;
  return PLAN_LIMITS[key] ?? PLAN_LIMITS.FREE;
}

export function progressBarColor(percent: number): string {
  if (percent >= 85) return "bg-red-500";
  if (percent >= 60) return "bg-amber-500";
  return "bg-emerald-500";
}

export function progressTrackColor(percent: number): string {
  if (percent >= 85) return "bg-red-500/15";
  if (percent >= 60) return "bg-amber-500/15";
  return "bg-emerald-500/15";
}
